paasta_tools/marathon_tools.py

```diff
@@ -269,21 +268,23 @@ def get_routing_constraints(self, service_namespace_config, system_paasta_config
         If, for example, a given app's 'discover' key is set to 'region', then this function
         computes the constraints required to group the app evenly amongst each
         of the actual 'region' values in the cluster.
-        It does so by querying the value of the region attribute for each slave
-        in the cluster, returning a GROUP_BY constraint where the value is the
-        number of unique regions.
+        It does so by querying the value of the discover attribute for each expected slave in the cluster (as defined
+        by the expected_slave_attributes key in system paasta config), returning a GROUP_BY constraint where the value
+        is the number of unique values for that attribute.
+        If you have not set expected_slave_attributes in the system paasta config, this function returns an empty list.
 
         :param service_namespace_config: the config for this service
         :returns: a set of constraints for marathon
         """
         discover_level = service_namespace_config.get_discover()
-        slaves = get_slaves()
-        if not slaves:
-            raise NoSlavesAvailableError(
-                "No slaves could be found in the cluster."
-            )
+
+        expected_slave_attributes = system_paasta_config.get_expected_slave_attributes()
+        if expected_slave_attributes is None:
+            return []
+
+        fake_slaves = [{"attributes": a} for a in expected_slave_attributes]
```

Comment:
Yep. And `filter_mesos_slaves_by_blacklist` is still used elsewhere, otherwise I would've refactored it to filter attributes.